src/graphql-schema.js

```diff
@@ -95,32 +96,33 @@ export const resolvers = {
       // if (user && user.id) {
       //   throw new Error('Already logged in.')
       // }
-
       const session = driver.session()
-      const res = await session.run('MATCH (u:User {email: "' + email + '"}) RETURN u.id, u.slug, u.name, u.avatar, u.email, u.password, u.role LIMIT 1')
-      let u = res.records[0]._fields ? zipObject([
-        'id',
-        'slug',
-        'name',
-        'avatar',
-        'email',
-        'password',
-        'role'
-      ], res.records[0]._fields) : null
-      if (u) {
-        if (await bcrypt.compareSync(password, u.password)) {
-          delete u.password
-          u.avatar = fixUrl(u.avatar)
-          return Object.assign(u, {
-            token: generateJwt(u)
-          })
+      try {
+        const res = await session.run('MATCH (u:User {email: "' + email + '"}) RETURN u.id, u.slug, u.name, u.avatar, u.email, u.password, u.role LIMIT 1')
+        const u = res.records[0]._fields ? zipObject([
+          'id',
+          'slug',
+          'name',
+          'avatar',
+          'email',
+          'password',
+          'role'
+        ], res.records[0]._fields) : null
+        if (u) {
+          if (await bcrypt.compareSync(password, u.password)) {
+            delete u.password
+            u.avatar = fixUrl(u.avatar)
+            return Object.assign(u, {
+              token: generateJwt(u)
+            })
+          }
+          session.close()
+          throw new AuthenticationError('Incorrect username or password.')
         }
         session.close()
-        throw new Error('Incorrect password.')
+      } catch (error) {
+        throw new AuthenticationError('Incorrect username or password.')
```

Comment:
```suggestion
        session.close()
        throw new AuthenticationError('Incorrect username or password.')
```

Reply:
that's a good question. i just went to the doc of the neo4j driver (https://www.npmjs.com/package/neo4j-driver) and they do not seem to close the session on error. they do it when the promise get resolved but they seem not to close it on error. i coudl add one but i can not tell if it is required or not.

Reply:
Hi, I have been lloking at what is done and i wanted to suggest to change the way we address login.  in fact I changed many things:

1. I remove usage of Lodash zipObejct as ES6 is doign that now,
2. I changed the cypher to use the key user so it is easier to query the result with a record.get
3. I used .then and closed the session at the beigining (so we are sure the session is closed whatever happens
4 . I am throwign error if user doe snot exist or if password is wrong in else statements.
I am not saying we shoudl use this way, but I think it is cleaner. 
I coudl created a new commit for this pull request but here is ho the new code would look like:
```      const session = driver.session()
      return session.run(
        'MATCH (user:User {email: "' + email + '"}) ' +
        'RETURN user {.id, .slug, .name, .avatar, .email, .password, .role} as user LIMIT 1')
        .then(async (result) => {
          session.close()
          const [currentUser] = await result.records.map(function (record) {
            console.log(record.get('user'))
            return record.get('user')
          })
          if (currentUser) {
            if (await bcrypt.compareSync(password, currentUser.password)) {
              delete currentUser.password
              currentUser.avatar = fixUrl(currentUser.avatar)
              return Object.assign(currentUser, {
                token: generateJwt(currentUser)
              })
            } else throw new AuthenticationError('Incorrect username or password.')
          } else throw new AuthenticationError('Incorrect username or password.')
        })```

Reply:
For readability we should never write cascaded code. (If Inside if inside then)

Try to keep max level of one, not three as the complexity rises very quick and becomes unmaintainable for others.

Reply:
agreed. i will provide a new version.

Reply:
a new commit with changes has been submittted